dockerUtils/dockerUtils.go

```diff
@@ -97,7 +100,86 @@ func (d *Docker) ContainerGetInfo(containerID string) (*ContainerInfo, error) {
 		return nil, err
 	}
 
+	rootfs := ""
+	if info.GraphDriver.Name == "overlay2" {
+		rootfs = info.GraphDriver.Data["MergedDir"]
+	}
+
 	return &ContainerInfo{
+		Rootfs:     rootfs,
 		AutoRemove: info.HostConfig.AutoRemove,
 	}, nil
 }
+
+// ContainerIsDocker returns true if the given container ID corresponds to a
+// Docker container. It does this by first trying to query Docker for the
+// container. If this doesn't work, it uses a heuristic based on the container's
+// rootfs.
+func ContainerIsDocker(id, rootfs string) (bool, error) {
+
+	timeout := time.Duration(500 * time.Millisecond)
+
+	docker, err := DockerConnect(timeout)
+	if err == nil {
+		defer docker.Disconnect()
+		_, err := docker.ContainerGetImageID(id)
+		return (err == nil), nil
+	}
+
+	// The connection to Docker can fail when containers are restarted
+	// automatically after reboot (i.e., containers originally launched with
+	// "--restart"); Docker won't respond until those are up. See Sysbox issue
+	// #184. In this case we determine if the container is a Docker container by
+	// examining the container's rootfs.
+
+	return isDockerRootfs(rootfs)
+}
+
+// isDockerRootfs determines if the given a container rootfs is for a Docker container.
+func isDockerRootfs(rootfs string) (bool, error) {
+
+	// Check if the container rootfs is under Docker's default data root
+	// (when in test-mode, we skip this so as to do the deeper check below)
+	if !testMode {
+		if strings.Contains(rootfs, "/var/lib/docker") {
+			return true, nil
+		}
+	}
+
+	// Check the parent dirs of the rootfs (up to 3 levels) and look for the
+	// `image, network, swarm, and containers` directories that are part of the
+	// Docker data root.
+
+	dockerDirs := []string{"image", "network", "containers", "swarm"}
+
+	searchLevels := 3
```

Comment:
If we already know the pattern utilized by docker to set the rootfs path (e.g. "/var/lib/docker/overlay2/l/<blah>"), why are you suggesting to iterate 3 levels up? I understand that docker allows you to configure the data-root, but AFAIK, there's no knob to set how the rootfs path looks like beyond the data-root suffix.

So why not avoiding this 3-level iteration and going straight to the expected path?

Reply:
That's only when using the overlay storage driver. I don't know what the behavior is with other storage drivers (e.g., aufs), so that's why I iterated up. In fact we should probably iterate more levels just to be sure.

Reply:
Got it.